File: searching/BinarySearch.java
package searching;

public class BinarySearch {

    public static void main(String[] args) {
	int[] arr = {1,2,3,4}; //{-3,2,5,8,22,67,89,101};
	System.out.println(binarySearchIterative(arr,1) == true);
	System.out.println(binarySearchIterative(arr,4) == true);
	System.out.println(binarySearchIterative(arr,2) == true);
	System.out.println(binarySearchIterative(arr,-3) == false);

	System.out.println(binarySearchRecursive(arr,1,0,arr.length) == 0);
	System.out.println(binarySearchRecursive(arr,4,0,arr.length) == 3);
	System.out.println(binarySearchRecursive(arr,2,0,arr.length) == 1);
	System.out.println(binarySearchRecursive(arr,-3,0,arr.length) == -1);

	int[] arr2 = {-3,2,5,8,22,67,89,101};
	System.out.println(binarySearchIterative(arr2,101) == true);
	System.out.println(binarySearchIterative(arr2,-3) == true);
	System.out.println(binarySearchIterative(arr2,8) == true);
	System.out.println(binarySearchIterative(arr2,99) == false);

	System.out.println(binarySearchRecursive(arr2,-3,0,arr2.length-1) == 0);
	System.out.println(binarySearchRecursive(arr2,101,0,arr2.length-1) == arr2.length-1);
	System.out.println(binarySearchRecursive(arr2,8,0,arr2.length-1) == 3);
	System.out.println(binarySearchRecursive(arr2,99,0,arr2.length-1) == -1);

    }

    /*
     * Return true if int is in Array
     * [1,3,5,6,7]
     * [1,2,3,4]
     */
    public static boolean binarySearchIterative(int[] arr, int target) {
	int low = 0;
	int high = arr.length;
	while (low < high) {
	    int mid = (high + low) / 2;
	    if (arr[mid] == target) {
		return true;
	    } else if (target < arr[mid]) {
		high = mid;
	    } else {
		low = mid+1;
	    }
	}
	return false;
    }

    /*
     * Return index if int is in Array, else -1
     * [1,3,5,6,7]
     * [1,2,3,4]
     */
    public static int binarySearchRecursive(int[] arr, int target, int low, int high) {
	int mid = (high + low) / 2;
	if (low > high) {
	    return -1;
	} else if (arr[mid] == target) {
	    return mid;
	} else if (target < arr[mid]) {
	    return binarySearchRecursive(arr, target, low, mid-1);
	} else {
	    return binarySearchRecursive(arr, target, mid+1, high);
	}
    }


}

File: searching/python/binary_search.py
"""
Binary Search

Write method to return True if integer is in array
"""


def binary_search(arr, val):
	low = 0
	high = len(arr)-1
	while low <= high:
		mid = low + (high-low)/2  #alternatively (high+low)/2
		if arr[mid] == val:
			return True
		elif arr[mid] < val:
			low = mid+1
		else:
			high = mid-1
	return False


def binary_search_recur(arr, val):
	if arr is None or len(arr) == 0:
		return False
	mid = len(arr)/2
	if arr[mid] == val:
		return True
	elif arr[mid] > val:
		return binary_search_recur(arr[:mid], val)
	else:
		return binary_search_recur(arr[mid+1:], val)




# Tests
a1 = [1,2,3,4,5]
a2 = [1,2]
a3 = [1]

def test_binary_search():
	assert binary_search(a1, 2) == True
	assert binary_search(a1, 9) == False
	assert binary_search(a1, 1) == True
	assert binary_search(a1, 0) == False
	
	assert binary_search(a2, 5) == False
	assert binary_search(a2, 1) == True
	assert binary_search(a2, 2) == True
	assert binary_search(a2, -3) == False

	assert binary_search(a3, 1) == True
	assert binary_search(a3, 0) == False
	assert binary_search(a3, 3) == False

def test_binary_search_recur():
	assert binary_search_recur(a1, 2) == True
	assert binary_search_recur(a1, 9) == False
	assert binary_search_recur(a1, 1) == True
	assert binary_search_recur(a1, 0) == False
	
	assert binary_search_recur(a2, 5) == False
	assert binary_search_recur(a2, 1) == True
	assert binary_search_recur(a2, 2) == True
	assert binary_search_recur(a2, -3) == False

	assert binary_search_recur(a3, 1) == True
	assert binary_search_recur(a3, 0) == False
	assert binary_search_recur(a3, 3) == False

if __name__ == "__main__":
	test_binary_search()
	test_binary_search_recur()

